rlba/environments/contextual_logistic_bandit.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2022 Morteza Ibrahimi. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A Contextual Logistic Bandit environment.
+"""
+
+
+import numpy as np
+from rlba.types import (
+    Array,
+    ArraySpec,
+    BoundedArraySpec,
+    DiscreteArraySpec,
+    NestedArraySpec,
+    NestedDiscreteArraySpec,
+)
+
+
+class ContextualLogisticBandit:
+    """This class creates a contextual logistic bandit environment."""
+
+    def __init__(self, num_actions, num_contexts, dim, seed, sigma_p=1):
+        self.num_actions = num_actions
+        self.num_contexts = num_contexts
+        self.dim = dim
+        self.sigma_p = sigma_p
+        self.sigma_p_squared = sigma_p**2
+        self.theta = sigma_p * np.random.randn(dim)
+
+        # Matrix of Phi_{S,A}
+        feature = np.random.randn(num_contexts, num_actions, dim)
+        self.feature = feature/np.linalg.norm(feature, axis=-1, keepdims=True)
+
+        exp_logits = np.exp(np.tensordot(self.feature,
+                                         self.theta, axes=([-1], [0])))
+
+        # Reward Probabilities for each (S,A)
+        self.rewards = exp_logits/(1 + exp_logits)
+        # Maximum Reward for each context S
+        self.values = np.max(self.rewards, axis=1, keepdims=True)
+        # Matrix of Regrets
+        self.regrets = self.values - self.rewards
+
+        self._action_spec = DiscreteArraySpec(num_actions, name="action spec")
+        self._observation_spec: ArraySpec = BoundedArraySpec(
+            shape=(2,),
+            dtype=int,
+            minimum=0,
+            maximum=num_contexts,
+            name="observation spec",
+        )
+        self._reset_context()
+        self._prev_context = None
+
+    def _reset_context(self):
+        """This function resets the context. For internal use only."""
+        self.context = np.random.randint(low=0, high=self.num_contexts)
+        self.context_features = self.feature[self.context, :, :]
+
+    def expected_reward(self, action):
```

Comment:
The default behavior of the bandit regret observer is now to cache previously seen actions. Please make sure in your use caching is disabled since this is a contextual bandit.